content/blog/2022/2022-08-03-feature-flags-in-nodejs-backend-server-with-launchdarkly.md

```diff
@@ -0,0 +1,720 @@
+
+Next we can define a similar kind of code that we did it earlier but we will subscribe to the log level changes before we initiate the `express` app:
+
+```javascript
+import util from 'util';
+import express from 'express';
+import LaunchDarkly from 'launchdarkly-node-server-sdk';
+import Logger from './logger.js';
+
+const PORT = 5000;
+const app = express();
+const simpleLogger = new Logger('Backend');
+
+const LD_SDK_KEY = 'sdk-d2432dc7-e56a-458b-9f93-0361af47d578';
+const LOG_LEVEL_FLAG_KEY = 'backend-log-level';
+const buyerKey = 'mob-b9d6d4d4-4300-46fa-9b13-d9eac89f9794';
+const client = LaunchDarkly.init(LD_SDK_KEY);
+client.waitForInitialization();
+const asyncGetFlag = util.promisify(client.variation);
+
+const subscribeToChanges = () => {
+  client.on(`update:${LOG_LEVEL_FLAG_KEY}`, (_, newValue) => {
+    const {
+      fallthrough: { variation },
+      variations
+    } = newValue;
+    const newLogLevel = variations[variation];
+    console.log(`${LOG_LEVEL_FLAG_KEY} updated to '${newLogLevel}'`);
+    Logger.setLogLevel(newLogLevel);
+  });
+};
+
+client.once('ready', async () => {
+  const user = { key: buyerKey, anonymous: true };
+  const initialLogLevel = await asyncGetFlag(LOG_LEVEL_FLAG_KEY, user, 'debug');
+  Logger.setLogLevel(initialLogLevel);
+
+  subscribeToChanges();
+
+  app.get('/', (req, res) => {
+    simpleLogger.debug('detailed debug message');
+    simpleLogger.log('simple log message');
+    simpleLogger.warn('Warning warning do something');
+    simpleLogger.error('ERROR! ERROR!');
+    res.sendStatus(200);
+  });
+
+  app.listen(PORT, () => {
+    simpleLogger.log(`Server listening on port ${PORT}`);
+  });
+});
+```
+
+Next we can define the script as part of `package.json`:
+
+```json
+{
+	"scripts": {
+    	"bootstrap": "node bootstrap.js"
+  	}
+}
+```
+
+Then we can execute the following command to run our app:
+
+```bash
+npm run bootstrap
+```
+
+Finally when we hit the API with the endpoint http://localhost:5000 we get to see the following log message being printed:
+
+```bash
+info: [LaunchDarkly] Initializing stream processor to receive feature flag updates
+info: [LaunchDarkly] Opened LaunchDarkly stream connection
+07-26-2022 11:54:58:193 Debug [Backend] detailed debug message
+07-26-2022 11:54:58:195 Log   [Backend] simple log message
+07-26-2022 11:54:58:196 Warn  [Backend] Warning warning do something
+07-26-2022 11:54:58:197 Error [Backend] ERROR! ERROR!
+```
+
+## Perform Admin Specific Operations using Multivariate Flags in LaunchDarkly
+
+LaunchDarkly supports something named *“multivariate”* flag apart from the simple boolean, String or JSON values. A multivariate feature flag could be a list of different strings, numbers or booleans. So let’s try defining few long-lived operational flag that would update our application feature dynamically.
+
+### Change Log Level on the Fly without Restarting the Server
```

Comment:
I'm not sure what the difference is between this section and the code in the bootstrapping section. Doesn't the boostrapping code already support exactly this? The code in this section seems like a more complicated version that does exactly the same thing?

Reply:
The previous one was a basic server startup code that would fetch the flags initially and load the server and host it. I thought of going with something basic and then make it more efficinet by converting it to fetch log levels dynamically. This covers the dynamic part.